flowmachine/flowmachine/core/subscriber_subset.py

```diff
@@ -0,0 +1,186 @@
+        )
+
+    @abstractmethod
+    def apply_subset(self, sql, *, subscriber_identifier):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement 'apply_subset'"
+        )
+
+    def _get_query_attrs_for_dependency_graph(self, analyse=False):
+        # This is a stub implementation of this internal method.
+        # It is needed because SubscriberSubsetBase currently
+        # inherits from flowmachine.Query, so we implement just
+        # enough to ensure Query.dependency_graph() doesn't break.
+        attrs = {}
+        attrs["name"] = self.__class__.__name__
+        attrs["stored"] = "N/A"
+        attrs["cost"] = "N/A"
+        attrs["runtime"] = "N/A"
+        return attrs
+
+
+class AllSubscribers(SubscriberSubsetBase):
+    """
+    Represents the subset of all subscribers (i.e., no subsetting at all).
+    """
+
+    is_proper_subset = False
+
+    def _make_query(self):
+        # Return a dummy string representing this subset. This is only needed
+        # because SubscriberSubsetBase currently inherits from Query, but will
+        # eventually be removed.
+        return "<AllSubscribers>"
+
+    def apply_subset(self, sql, *, subscriber_identifier=None):
+        """
+        Return the input query unchanged, since no subsetting is applied.
+
+        Parameters
+        ----------
+        sql : sqlalchemy.sql.ClauseElement
+            The SQL query to which the subset should be applied.
+
+        subscriber_identifier : str
+            This argument is ignored for subsets of type 'AllSubscribers'.
+        """
+        return sql
+
+
+class SubsetFromFlowmachineQuery(SubscriberSubsetBase):
+    """
+    Represents a subset given by a flowmachine query.
+    """
+
+    is_proper_subset = True
+
+    def __init__(self, flowmachine_query):
+        assert isinstance(flowmachine_query, Query)
+
+        self.subset_query = text(flowmachine_query.get_query()).columns(
+            column("subscriber")
+        )
+
+    def _make_query(self):
+        # Return a dummy string representing this subset. This is only needed
+        # because SubscriberSubsetBase currently inherits from Query, but will
+        # eventually be removed.
+        return "<SubsetFromFlowmachineQuery>"
+
+    def apply_subset(self, sql, *, subscriber_identifier=None):
+        """
+        Return a modified version of the input SQL query which has the subset applied.
+
+        Parameters
+        ----------
+        sql : sqlalchemy.sql.ClauseElement
+            The SQL query to which the subset should be applied.
+
+        subscriber_identifier : str
+            This argument is ignored for subsets of type 'SubsetFromFlowmachineQuery'.
+
+        Returns
+        ----------
+        sqlalchemy.sql.ClauseElement
+        """
+        assert isinstance(sql, ClauseElement)
+
+        tbl = sql.alias("tbl")
+
+        try:
+            subset = self.subset_query.distinct().alias("subset")
```

Comment:
Didn't you use `text(flowmachine.get_query())` to transform the query into an sqlalchemy query? So when would we get an error here?

Reply:
Yes, you are right. Or rather, I think this will _always_ raise an error because the textual query doesn't have a `.distinct()` attribute. I think I was confused when I wrote this. 😂Should be clearer once I've reworked it following your suggestion above to move `flowmachine.get_query()` down here.